File: src/Driver.Domain/Entities/Incident.cs
using DriverDms.Domain.Common;

namespace DriverDms.Domain.Entities;

public class Incident : BaseEntity
{
    public int IncidentId { get; set; }
    public int DriverId { get; set; }
    public Driver? Driver { get; set; }
    public DateTime IncidentDate { get; set; }
    public int IncidentTypeId { get; set; }
    public IncidentType? IncidentType { get; set; }
    public string Description { get; set; } = default!;
    public int SeverityLevelId { get; set; }
    public SeverityLevel? SeverityLevel { get; set; }
    public string? Location { get; set; }
    public bool RootCauseCompleted { get; set; }

    // Captured at report time from the driver's current transporter (auto-filled on
    // the frontend), or overridden with the newly reported one when TransporterChanged
    // is true. Purely informational - does not itself alter DriverTransporterHistory.
    public string? TransporterName { get; set; }
    public bool TransporterChanged { get; set; }

    // Relative path only (e.g. "incident-reports/<guid>.pdf") - see Driver.DriverPhotoPath note.
    public string? IncidentReportPath { get; set; }

    public ICollection<CorrectiveAction> CorrectiveActions { get; set; } = new List<CorrectiveAction>();
}

public class CorrectiveAction : BaseEntity
{
    public int CorrectiveActionId { get; set; }
    public int IncidentId { get; set; }
    public Incident? Incident { get; set; }
    public string ActionTaken { get; set; } = default!;
    public int? PenaltyTypeId { get; set; }
    public PenaltyType? PenaltyType { get; set; }
    public DateTime ActionDate { get; set; }

    // Relative path only (e.g. "corrective-actions/<guid>.pdf") - see Driver.DriverPhotoPath
    // note. Used for the letter-of-apology / supporting-photo attachment on a corrective
    // action. Required (alongside HOD/Admin review) before a Blacklisted driver can be
    // unblocked - see DriverService.UpdateStatusAsync.
    public string? AttachmentPath { get; set; }

    // Relative path only (e.g. "incident-reports/{incidentId}/corrective-actions/<guid>.pdf") -
    // see Driver.DriverPhotoPath note. Stores the formal apology document uploaded when
    // a corrective action is recorded. Namespaced under the parent incident folder so all
    // apology documents for a given incident are co-located on disk without a DB lookup.
    public string? ApologyDocumentPath { get; set; }
}

File: dms-frontend/src/types/incident.ts
export interface Incident {
  incidentId: number;
  driverId: number;
  driverName: string;
  incidentTypeId: number;
  incidentTypeName: string;
  incidentDate: string;
  location?: string;
  description?: string;
  severity: string;
  faultAtDriver: boolean;
  fatalitiesCount: number;
  injuriesCount: number;
  correctiveActions: CorrectiveAction[];
  /** True once a report has been uploaded via POST /api/incidents/{id}/report */
  hasReport: boolean;
  /** Transporter on file for the driver at the time of this incident */
  transporterName?: string | null;
  /** True if the employee reported the driver had since moved to a different contractor */
  transporterChanged: boolean;
}

export interface CorrectiveAction {
  correctiveActionId: number;
  incidentId: number;
  actionTaken: string;
  penaltyTypeName?: string | null;
  actionDate: string;
  /** True once a letter of apology / supporting photo has been uploaded */
  hasAttachment: boolean;
  /** True once a formal apology document was uploaded at corrective-action creation time */
  hasApologyDocument: boolean;
}

export interface IncidentSearchParams {
  page?: number;
  pageSize?: number;
  driverId?: number;
  severity?: string;
  dateFrom?: string;
  dateTo?: string;
  /** Filters to incidents whose driver holds a license number containing this text. */
  licenseNo?: string;
}

export interface CreateIncidentRequest {
  driverId: number;
  incidentTypeId: number;
  incidentDate: string;
  location?: string;
  description?: string;
  severity: string;
  faultAtDriver: boolean;
  fatalitiesCount: number;
  injuriesCount: number;
  /** Auto-filled from the driver's current transporter, or the newly reported one if transporterChanged is true */
  transporterName?: string | null;
  transporterChanged: boolean;
}

export type UpdateIncidentRequest = Partial<Omit<CreateIncidentRequest, "driverId">>;

export interface CreateCorrectiveActionRequest {
  incidentId: number;
  actionTaken: string;
  penaltyTypeId?: number;
  actionDate: string;
}

/** Alias used by incidentService */
export type AddCorrectiveActionRequest = CreateCorrectiveActionRequest;

